File: src/C3.Infrastructure/CatalogueFiles/Xml/V1_1/LegacyBrandRepository.vb
Imports C3.Catalogue.Brands
Imports System.Data

Namespace CatalogueFiles.Xml.V1_1

    Public NotInheritable Class LegacyBrandRepository
        Implements IBrandRepository

        Private ReadOnly _documentProvider As Func(Of DataSet)

        Public Sub New(documentProvider As Func(Of DataSet))
            If documentProvider Is Nothing Then
                Throw New ArgumentNullException("documentProvider")
            End If
            _documentProvider = documentProvider
        End Sub

        Public Function GetAll() As IList(Of Brand) Implements IBrandRepository.GetAll
            Dim values As New List(Of Brand)()
            For Each row As DataRow In BrandsTable().Rows
                If row.RowState <> DataRowState.Deleted Then
                    values.Add(Map(row))
                End If
            Next
            Return values
        End Function

        Public Function FindByCode(code As String) As Brand Implements IBrandRepository.FindByCode
            If String.IsNullOrWhiteSpace(code) Then
                Return Nothing
            End If

            For Each row As DataRow In BrandsTable().Rows
                If row.RowState <> DataRowState.Deleted AndAlso
                        String.Equals(CStr(row("Code")), code, StringComparison.OrdinalIgnoreCase) Then
                    Return Map(row)
                End If
            Next
            Return Nothing
        End Function

        Public Function IsCodeInUse(code As String) As Boolean Implements IBrandRepository.IsCodeInUse
            Return FindByCode(code) IsNot Nothing
        End Function

        Public Function IsReferencedByModel(code As String) As Boolean _
                Implements IBrandRepository.IsReferencedByModel

            Dim brandRow As DataRow = FindRow(code)
            Dim legacyName As String = String.Empty
            If brandRow IsNot Nothing Then
                legacyName = Convert.ToString(brandRow("Brand"))
            End If

            For Each row As DataRow In ModelsTable().Rows
                If row.RowState <> DataRowState.Deleted Then
                    Dim storedBrand As String = Convert.ToString(row("Brand"))
                    If String.Equals(storedBrand, code, StringComparison.OrdinalIgnoreCase) OrElse
                            (legacyName.Length > 0 AndAlso
                             String.Equals(storedBrand, legacyName, StringComparison.OrdinalIgnoreCase)) Then
                        Return True
                    End If
                End If
            Next
            Return False
        End Function

        Public Sub Add(value As Brand) Implements IBrandRepository.Add
            BrandsTable().Rows.Add(value.Name, value.Code, value.AddedAt, value.Notes)
            SynchronizeBrandCounter()
        End Sub

        Public Sub Update(value As Brand) Implements IBrandRepository.Update
            Dim row As DataRow = FindRow(value.Code)
            If row Is Nothing Then
                Throw New InvalidOperationException("The selected brand no longer exists.")
            End If

            Dim previousName As String = Convert.ToString(row("Brand"))
            row("Brand") = value.Name
            row("Date") = value.AddedAt
            row("Notes") = value.Notes

            ' Older C3 files stored the display name in Models.Brand. Migrate
            ' those references to the stable code when the owning brand changes.
            For Each modelRow As DataRow In ModelsTable().Rows
                If modelRow.RowState <> DataRowState.Deleted AndAlso
                        String.Equals(
                            Convert.ToString(modelRow("Brand")),
                            previousName,
                            StringComparison.OrdinalIgnoreCase) Then
                    modelRow("Brand") = value.Code
                End If
            Next
        End Sub

        Public Sub Delete(code As String) Implements IBrandRepository.Delete
            Dim row As DataRow = FindRow(code)
            If row Is Nothing Then
                Throw New InvalidOperationException("The selected brand no longer exists.")
            End If

            BrandsTable().Rows.Remove(row)
            SynchronizeBrandCounter()
        End Sub

        Private Function Document() As DataSet
            Dim value As DataSet = _documentProvider()
            If value Is Nothing Then
                Throw New InvalidOperationException("No active catalogue document is available.")
            End If
            Return value
        End Function

        Private Function BrandsTable() As DataTable
            Return RequireTable("Brands")
        End Function

        Private Function ModelsTable() As DataTable
            Return RequireTable("Models")
        End Function

        Private Function RequireTable(name As String) As DataTable
            Dim table As DataTable = Document().Tables(name)
            If table Is Nothing Then
                Throw New InvalidOperationException("Catalogue table '" & name & "' is missing.")
            End If
            Return table
        End Function

        Private Function FindRow(code As String) As DataRow
            For Each row As DataRow In BrandsTable().Rows
                If row.RowState <> DataRowState.Deleted AndAlso
                        String.Equals(CStr(row("Code")), code, StringComparison.OrdinalIgnoreCase) Then
                    Return row
                End If
            Next
            Return Nothing
        End Function

        Private Shared Function Map(row As DataRow) As Brand
            Dim addedAt As DateTime = DateTime.MinValue
            If Not row.IsNull("Date") Then
                addedAt = Convert.ToDateTime(row("Date"))
            End If

            Return New Brand(
                Convert.ToString(row("Brand")),
                Convert.ToString(row("Code")),
                addedAt,
                Convert.ToString(row("Notes")))
        End Function

        Private Sub SynchronizeBrandCounter()
            Dim counters As DataTable = RequireTable("Counters")
            Dim counter As DataRow = counters.Rows.Find("Brands")
            If counter IsNot Nothing Then
                counter("Number") = BrandsTable().Rows.Count
            End If
        End Sub

    End Class

End Namespace

File: src/C3.Infrastructure/CatalogueFiles/Xml/V1_1/LegacyBrandRepository.cs
using C3.Catalogue.Brands;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Runtime.CompilerServices;

namespace C3.Infrastructure.CatalogueFiles.Xml.V1_1
{
	public sealed class LegacyBrandRepository : IBrandRepository
	{
		private readonly Func<DataSet> _documentProvider;

		public LegacyBrandRepository(Func<DataSet> documentProvider)
		{
			if (documentProvider == null)
			{
				throw new ArgumentNullException("documentProvider");
			}
			_documentProvider = documentProvider;
		}

		public IList<Brand> GetAll()
		{
			List<Brand> list = new List<Brand>();
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = BrandsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted)
					{
						list.Add(Map(dataRow));
					}
				}
				return list;
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
		}

		IList<Brand> IBrandRepository.GetAll()
		{
			//ILSpy generated this explicit interface implementation from .override directive in GetAll
			return this.GetAll();
		}

		public Brand FindByCode(string code)
		{
			if (string.IsNullOrWhiteSpace(code))
			{
				return null;
			}
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = BrandsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted && string.Equals(Conversions.ToString(dataRow["Code"]), code, StringComparison.OrdinalIgnoreCase))
					{
						return Map(dataRow);
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return null;
		}

		Brand IBrandRepository.FindByCode(string code)
		{
			//ILSpy generated this explicit interface implementation from .override directive in FindByCode
			return this.FindByCode(code);
		}

		public bool IsCodeInUse(string code)
		{
			return FindByCode(code) != null;
		}

		bool IBrandRepository.IsCodeInUse(string code)
		{
			//ILSpy generated this explicit interface implementation from .override directive in IsCodeInUse
			return this.IsCodeInUse(code);
		}

		public bool IsReferencedByModel(string code)
		{
			DataRow dataRow = FindRow(code);
			string text = string.Empty;
			if (dataRow != null)
			{
				text = Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Brand"]));
			}
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = ModelsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow2 = (DataRow)enumerator.Current;
					if (dataRow2.RowState != DataRowState.Deleted)
					{
						string a = Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow2["Brand"]));
						if (!string.Equals(a, code, StringComparison.OrdinalIgnoreCase) && (text.Length <= 0 || !string.Equals(a, text, StringComparison.OrdinalIgnoreCase)))
						{
							continue;
						}
						return true;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return false;
		}

		bool IBrandRepository.IsReferencedByModel(string code)
		{
			//ILSpy generated this explicit interface implementation from .override directive in IsReferencedByModel
			return this.IsReferencedByModel(code);
		}

		public void Add(Brand value)
		{
			BrandsTable().Rows.Add(value.Name, value.Code, value.AddedAt, value.Notes);
			SynchronizeBrandCounter();
		}

		void IBrandRepository.Add(Brand value)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Add
			this.Add(value);
		}

		public void Update(Brand value)
		{
			DataRow dataRow = FindRow(value.Code);
			if (dataRow == null)
			{
				throw new InvalidOperationException("The selected brand no longer exists.");
			}
			string b = Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Brand"]));
			dataRow["Brand"] = value.Name;
			dataRow["Date"] = value.AddedAt;
			dataRow["Notes"] = value.Notes;
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = ModelsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow2 = (DataRow)enumerator.Current;
					if (dataRow2.RowState != DataRowState.Deleted && string.Equals(Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow2["Brand"])), b, StringComparison.OrdinalIgnoreCase))
					{
						dataRow2["Brand"] = value.Code;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
		}

		void IBrandRepository.Update(Brand value)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Update
			this.Update(value);
		}

		public void Delete(string code)
		{
			DataRow dataRow = FindRow(code);
			if (dataRow == null)
			{
				throw new InvalidOperationException("The selected brand no longer exists.");
			}
			BrandsTable().Rows.Remove(dataRow);
			SynchronizeBrandCounter();
		}

		void IBrandRepository.Delete(string code)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Delete
			this.Delete(code);
		}

		private DataSet Document()
		{
			DataSet dataSet = _documentProvider();
			if (dataSet == null)
			{
				throw new InvalidOperationException("No active catalogue document is available.");
			}
			return dataSet;
		}

		private DataTable BrandsTable()
		{
			return RequireTable("Brands");
		}

		private DataTable ModelsTable()
		{
			return RequireTable("Models");
		}

		private DataTable RequireTable(string name)
		{
			DataTable dataTable = Document().Tables[name];
			if (dataTable == null)
			{
				throw new InvalidOperationException("Catalogue table '" + name + "' is missing.");
			}
			return dataTable;
		}

		private DataRow FindRow(string code)
		{
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = BrandsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted && string.Equals(Conversions.ToString(dataRow["Code"]), code, StringComparison.OrdinalIgnoreCase))
					{
						return dataRow;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return null;
		}

		private static Brand Map(DataRow row)
		{
			DateTime addedAt = DateTime.MinValue;
			if (!row.IsNull("Date"))
			{
				addedAt = Convert.ToDateTime(RuntimeHelpers.GetObjectValue(row["Date"]));
			}
			return new Brand(Convert.ToString(RuntimeHelpers.GetObjectValue(row["Brand"])), Convert.ToString(RuntimeHelpers.GetObjectValue(row["Code"])), addedAt, Convert.ToString(RuntimeHelpers.GetObjectValue(row["Notes"])));
		}

		private void SynchronizeBrandCounter()
		{
			DataRow dataRow = RequireTable("Counters").Rows.Find("Brands");
			if (dataRow != null)
			{
				dataRow["Number"] = BrandsTable().Rows.Count;
			}
		}
	}
}
